### PR TITLE
AA-475: Initial commit for the ERC-7562 validation rules native tracer

### PR DESCRIPTION
**Pull request by forshtat**

### CODE REVIEW

File: eth/tracers/native/erc7562.go [hunk 1]
```diff
@@ -0,0 +1,274 @@
...
+		opcode == "CALLCODE" ||
+		opcode == "DELEGATECALL" ||
+		opcode == "STATICCALL"
+}
+
+// not using 'isPrecompiled' to only allow the ones defined by the ERC-7562 as stateless precompiles
+// [OP-062]
+func isAllowedPrecompile(addr common.Address) bool {
+	addrInt := addr.Big()
+	return addrInt.Cmp(big.NewInt(0)) == 1 && addrInt.Cmp(big.NewInt(10)) == -1
```
Comment (drortirosh) [inline eth/tracers/native/erc7562.go:266]: Add constant list (could bit a bitmap) of precompiles. we already know if one more 0x100, defined in eip-7212

File: eth/tracers/native/erc7562.go [hunk 2]
```diff
@@ -0,0 +1,274 @@
...
+	tracers.DefaultDirectory.Register("callTracerWithOpcodes", newCallTracerWithOpcodes, false)
+}
+
+type callFrameWithOpcodes struct {
+	callFrame
+	AccessedSlots     accessedSlots          `json:"accessedSlots"`
+	ExtCodeAccessInfo []common.Address       `json:"extCodeAccessInfo"`
+	DeployedContracts []common.Address       `json:"deployedContracts"`
+	UsedOpcodes       map[string]bool        `json:"usedOpcodes"`
+	GasObserved       bool                   `json:"gasObserved"`
```
Comment (drortirosh) [inline eth/tracers/native/erc7562.go:45]: This is a new mechanism, trying to detect "gas/call" sequence, which doesn't use the stadnard "delay by 1" opcode check.
There is at least one more case we need such delay mechanism: EXTCODESIZE access to entrypoint is not allowed. But it is OK in the sequence of EXTCODESIZE+ISZERO

File: eth/tracers/native/erc7562.go [hunk 3]
```diff
@@ -0,0 +1,274 @@
...
+
+func init() {
+	tracers.DefaultDirectory.Register("callTracerWithOpcodes", newCallTracerWithOpcodes, false)
+}
+
+type callFrameWithOpcodes struct {
+	callFrame
+	AccessedSlots     accessedSlots          `json:"accessedSlots"`
+	ExtCodeAccessInfo []common.Address       `json:"extCodeAccessInfo"`
+	DeployedContracts []common.Address       `json:"deployedContracts"`
```
Comment (drortirosh) [inline eth/tracers/native/erc7562.go:43]: when  do we need deployed ?

File: eth/tracers/native/erc7562.go [hunk 4]
```diff
@@ -0,0 +1,274 @@
...
+}
+
+type callTracerWithOpcodes struct {
+	callTracer
+	env *tracing.VMContext
+
+	allowedOpcodeRegex *regexp.Regexp
+	lastOp             string
+	callstack          []callFrameWithOpcodes
+	lastThreeOpCodes   []string
```
Comment (drortirosh) [inline eth/tracers/native/erc7562.go:66]: remove last3ops. was never really in use (it was a placeholder for much more complex op-sequence check)

File: eth/tracers/native/erc7562.go [hunk 5]
```diff
@@ -0,0 +1,274 @@
...
+	DeployedContracts []common.Address       `json:"deployedContracts"`
+	UsedOpcodes       map[string]bool        `json:"usedOpcodes"`
+	GasObserved       bool                   `json:"gasObserved"`
+	ContractSize      map[common.Address]int `json:"contractSize"`
+	OutOfGas          bool                   `json:"outOfGas"`
+	Calls             []callFrameWithOpcodes `json:"calls,omitempty" rlp:"optional"`
+}
+
+// TODO: I suggest that we provide an `[]string` for all of these fields. Doing it for `Reads` as an example here.
+type accessedSlots struct {
```
[the remark this answers is not in the document]
Comment (forshtat) [reply eth/tracers/native/erc7562.go:52]: Bit for a bitmap? Why would I need a `uint64` for that?

Basically, options of what we can return here:
1. Last read/written value
2. All read/written values
3. Array of opcodes used for the slot
4. Bitmap of opcodes used for the slot (same semantic, save a few bits)
5. Counter of reads/writes
6. The 🤡 emoji
7. A single boolean if the slot was either read from or written to
8. A combination of the above

Please advise what is the best return value for this structure. Thanks!

File: eth/tracers/native/erc7562.go [hunk 6]
```diff
@@ -0,0 +1,274 @@
...
+type callFrameWithOpcodes struct {
+	callFrame
+	AccessedSlots     accessedSlots          `json:"accessedSlots"`
+	ExtCodeAccessInfo []common.Address       `json:"extCodeAccessInfo"`
+	DeployedContracts []common.Address       `json:"deployedContracts"`
+	UsedOpcodes       map[string]bool        `json:"usedOpcodes"`
+	GasObserved       bool                   `json:"gasObserved"`
+	ContractSize      map[common.Address]int `json:"contractSize"`
+	OutOfGas          bool                   `json:"outOfGas"`
+	Calls             []callFrameWithOpcodes `json:"calls,omitempty" rlp:"optional"`
```
[the remark this answers is not in the document]
Comment (forshtat) [reply eth/tracers/native/erc7562.go:48]: But the new type is inheriting the old type, therefore any non type safe code would have no issues parsing the return value with extra fields. The shape would not fit though, but this is in fact a different tracer so no reason to expect the same response shape.
There is also no reason to parse JSON into a golang `callFrameWithOpcodes` that I can think of right now.

Should I split up the structure into a base one and an extra struct we are adding? I can do it, it is not a problem.

**Comment by forshtat** [comment]
Actually, work will continue in https://github.com/eth-infinitism/go-ethereum/pull/47 and ported to the `ethereum/master` branch later

File: eth/tracers/native/erc7562.go [hunk 7]
```diff
@@ -0,0 +1,274 @@
...
+		call.Gas = t.gasLimit
+	}
+	t.callstack = append(t.callstack, call)
+}
+
+// TODO: Why is this type even needed?
+type partialStack = []*uint256.Int
+
+func (t *callTracerWithOpcodes) OnOpcode(pc uint64, op byte, gas, cost uint64, scope tracing.OpContext, rData []byte, depth int, err error) {
+	opcode := vm.OpCode(op).String()
```
[the remark this answers is not in the document]
Comment (forshtat) [reply eth/tracers/native/erc7562.go:138]: I started removing the `String()` conversion and all it but `handleExtOpcodes` relies on strings and I am not rewriting it yet. Added it in TODO, will fix soon.

File: eth/tracers/native/erc7562.go [hunk 8]
```diff
@@ -0,0 +1,274 @@
...
+}
+
+func (t *callTracerWithOpcodes) storeKeccak(opcode string, scope tracing.OpContext) {
+	if opcode == "KECCAK256" {
+		dataOffset := peepStack(scope.StackData(), 0).Uint64()
+		dataLength := peepStack(scope.StackData(), 1).Uint64()
+		memory := scope.MemoryData()
+		keccak := make([]byte, dataLength)
+		copy(keccak, memory[dataOffset:dataOffset+dataLength])
+		t.Keccak = append(t.Keccak, keccak)
```
[the remark this answers is not in the document]
Comment (forshtat) [reply eth/tracers/native/erc7562.go:209]: >not needed per-callframe, only unique values
Please elaborate, why not? Can a Paymaster access a storage slot that is associated with the Factory contract? This seems like artificially limiting the tracer for no reason.

File: eth/tracers/native/erc7562.go [hunk 9]
```diff
@@ -0,0 +1,274 @@
...
+	}
+}
+
+func (t *callTracerWithOpcodes) detectOutOfGas(gas uint64, cost uint64, opcode string, currentCallFrame callFrameWithOpcodes) {
```
[the remark this answers is not in the document]
Comment (forshtat) [reply eth/tracers/native/erc7562.go:213]: I don't see it in the code, how does it look like? The `callFrame.error` is just an error object, what does it contain?

File: eth/tracers/native/erc7562.go (hunk 2 above)
Comment (forshtat) [reply eth/tracers/native/erc7562.go:45]: The point is that the "opcode count" structure ignoring the `GAS` opcode if it was not "observed" is very weird and confusing. Extracting it into a separate field makes much more sense to me.

File: eth/tracers/native/erc7562.go (hunk 3 above)
Comment (forshtat) [reply eth/tracers/native/erc7562.go:43]: For the factory creation of the sender check. It is a generally safer and more informative then "allow single CREATE2 opcode" approach we had before.

File: eth/tracers/native/erc7562.go (hunk 4 above)
Comment (forshtat) [reply eth/tracers/native/erc7562.go:66]: How do I check the "gas observed" or "code size observed" conditions without `lastThreeOpCodes`?

File: eth/tracers/native/erc7562.go [hunk 10]
```diff
@@ -0,0 +1,274 @@
...
+// newCallTracer returns a native go tracer which tracks
+// call frames of a tx, and implements vm.EVMLogger.
+func newCallTracerWithOpcodes(ctx *tracers.Context, cfg json.RawMessage, chainConfig *params.ChainConfig) (*tracers.Tracer, error) {
+	t, err := newCallTracerObjectWithOpcodes(ctx, cfg)
+	if err != nil {
+		return nil, err
+	}
+	return &tracers.Tracer{
+		Hooks: &tracing.Hooks{
+			OnTxStart: t.OnTxStart,
```
[the remark this answers is not in the document]
Comment (forshtat) [reply eth/tracers/native/erc7562.go:79]: Fixed.

**Pull request closed without merging**